### ISSUE TITLE
Improve BlockWindow performance

### ISSUE THREAD

**Issue by stasatdaglabs** [kaspanet/kaspad#1691]
BlockWindow plays a role in calculating pastMedianTime and DAA.
The current algorithm is the most computationally intensive part of adding a block to the DAG beside transaction signature validation.

To resolve this issue, please design and implement a solution that will either improve BlockWindow or remove the dependency on it altogether.

**Comment by zawy12** [comment]
If there is a target DAG width, or it you know the current DAG width, then use median timestamp & target of the block's parents as the input to BCH's new ASERT.  This only requires looking at the block's parents.

**Comment by hashdag** [comment]
How about using the timestamp of the earliest chain block in the window? That is,
```
T = targetTimePerBlock
N = windowSize
B = minimal chain block in the window
target = avgTarget * ((current.Timestamp - B.Timestamp) / (T * N))
```
(I'm not sure what `avgTarget' stands for here, I'm following the description by @zawy12 above)

This seems close enough to the current design (security wise), and can significantly simplify the implementation and remove the overhead on performance, since you only need to keep track of each block's DAA predecessor (can be done similarly to / inside the code that tracks a block's finality point)

**Comment by someone235** [comment]
Fixed by #1948
